src/poetry/utils/authenticator.py

```diff
@@ -259,21 +262,29 @@ def request(
                 if is_last_attempt:
                     raise e
             else:
-                if resp.status_code not in [502, 503, 504] or is_last_attempt:
+                if resp.status_code not in STATUS_FORCELIST or is_last_attempt:
                     if raise_for_status:
                         resp.raise_for_status()
                     return resp
 
             if not is_last_attempt:
                 attempt += 1
-                delay = 0.5 * attempt
+                delay = self._get_backoff(resp, attempt)
                 logger.debug("Retrying HTTP request in %s seconds.", delay)
                 time.sleep(delay)
                 continue
 
         # this should never really be hit under any sane circumstance
         raise PoetryException("Failed HTTP {} request", method.upper())
 
+    def _get_backoff(self, response: requests.Response | None, attempt: int) -> float:
+        if response is not None:
+            retry_after = response.headers.get(RETRY_AFTER_HEADER, "")
+            if retry_after:
+                return float(retry_after)
+
+        return 0.5 * attempt
```

Comment:
mypy wasn't too happy about my try/except and duck typing here, hence the if/else checks. Happy to change this though.

Reply:
Should be fine as is.